### PR TITLE
Fix cfssl and etcd provisioners

### PR DESCRIPTION
**Pull request by ffilippopoulos**

### CODE REVIEW

File: cfssl.tf [hunk 1]
```diff
@@ -98,6 +76,27 @@ resource "aws_volume_attachment" "cfssl-data" {
   device_name = "/dev/xvdf"
   volume_id   = "${aws_ebs_volume.cfssl-data.*.id[count.index]}"
   instance_id = "${aws_instance.cfssl.*.id[count.index]}"
+
+  provisioner "remote-exec" {
+    when = "destroy"
+
+    inline = [
+      "sudo systemctl stop cfssl.service",
+      "sudo umount -d /dev/xvdf",
```
Comment (ffilippopoulos) [inline cfssl.tf]: changed to match the docs https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/ebs-detaching-volume.html

File: cfssl.tf [hunk 2]
```diff
@@ -98,6 +76,27 @@ resource "aws_volume_attachment" "cfssl-data" {
...
+      "sudo reboot",
```
Comment (ffilippopoulos) [inline cfssl.tf]: The above 2 commands were not enough to get rid of all the dependencies and aws was still failing on volume detach without restarting the instance first.

File: etcd.tf [hunk 3]
```diff
@@ -103,6 +83,27 @@ resource "aws_volume_attachment" "etcd-data" {
   device_name = "/dev/xvdf"
   volume_id   = "${aws_ebs_volume.etcd-data.*.id[count.index]}"
   instance_id = "${aws_instance.etcd.*.id[count.index]}"
+
+  provisioner "remote-exec" {
+    when = "destroy"
+
+    inline = [
```
Comment (ffilippopoulos) [inline etcd.tf]: same as above

File: cfssl.tf (hunk 1 above)
Comment (george-angel) [inline cfssl.tf]: Can you use `${self.device_name}` here?

File: cfssl.tf (hunk 2 above)
Comment (george-angel) [reply cfssl.tf]: Why `reboot`, you want a new instance, something like `halt` or `shutdown` would make more sense?

File: cfssl.tf [hunk 4]
```diff
@@ -98,6 +76,27 @@ resource "aws_volume_attachment" "cfssl-data" {
...
+    ]
+
+    connection {
+      bastion_host = "jumpbox.dev.uw.systems"
```
Comment (george-angel) [inline cfssl.tf]: This will only be true for dev, but fail in prod.

File: cfssl.tf [hunk 5]
```diff
@@ -98,6 +76,27 @@ resource "aws_volume_attachment" "cfssl-data" {
...
+    when = "destroy"
+
+    inline = [
+      "sudo systemctl stop cfssl.service",
+      "sudo umount -d ${self.device_name}",
+      "sudo halt",
+    ]
+
+    connection {
+      bastion_host = "jumpbox.${var.environment}.uw.systems"
```
Comment (george-angel) [inline cfssl.tf]: We should rethink this in a bigger way. Afaik this is the first change that introduces a hard link to UW. Before this it *could* be a generic module.

The logic to execute commands on the host is "generic" workaround of Terraform limitation. But the rest is completely tied to the way we operate. Module should let you provide a way to run the provision.

Comment (ffilippopoulos) [reply cfssl.tf]: Ah, that's true. But the jumpboxes are really something that we use but cannot require it by anyone who wants to use this module. I somehow need to make these optional I guess

**Comment by ffilippopoulos** [comment]
Updated to just using skip_destroy under aws_volume_attachments

**Review by george-angel — APPROVED** [review]

**Pull request merged**